xla/tests/hlo_test_base.cc

```diff
@@ -500,6 +542,120 @@ ::testing::AssertionResult HloTestBase::RunAndCompare(
+
+  for (auto [expected, actual] : llvm::zip_equal(test_0, test_1)) {
+    auto compare_result = LiteralTestUtil::NearOrEqual(expected, actual, error);
+    if (!compare_result) {
+      return compare_result;
+    }
+  }
+  return ::testing::AssertionSuccess();
+}
+
+::testing::AssertionResult HloTestBase::RunAndCompareTwoModulesReplicated(
+    std::unique_ptr<HloModule> module_0, std::unique_ptr<HloModule> module_1,
+    HloRunner::ReplicatedExecuteOptions options,
+    const optional<ErrorSpec>& error) {
+  int replica_count = module_0->config().replica_count();
+  if (replica_count != module_1->config().replica_count()) {
+    return ::testing::AssertionFailure()
+           << "Number of replicas is not the same: " << replica_count << " Vs "
+           << module_1->config().replica_count();
+  }
+  if (options.num_replicas != replica_count) {
+    return ::testing::AssertionFailure()
+           << "Number of execution replicas is different from number of "
+              "replicas in the module: requested number of replicas = "
+           << options.num_replicas
+           << ", number of replicas in hlo = " << replica_count;
+  }
+
+  std::vector<int> mismatches = CompareInputs(*module_0, *module_1);
+  if (!mismatches.empty()) {
+    return ::testing::AssertionFailure()
+           << "Error: parameter mismatch at indices: "
+           << absl::StrJoin(mismatches, ",");
+  }
+  auto num_args = module_0->entry_computation()->num_parameters();
+  if (num_args != options.arguments.size()) {
+    return ::testing::AssertionFailure()
+           << "Mismatch in number of arguments passed while running replicated "
+              "hlo module. Expected: "
+           << num_args << ", actual: " << options.arguments.size();
+  }
+  auto result = RunAndCompareTwoModulesInternalReplicated(
+      std::move(module_0), std::move(module_1), options, error);
+  if (!result.ok()) {
+    return ::testing::AssertionFailure() << result.status();
+  }
+  return result.value();
+}
+
+::testing::AssertionResult HloTestBase::RunAndCompareTwoModulesReplicated(
+    std::unique_ptr<HloModule> module_0, std::unique_ptr<HloModule> module_1,
+    bool run_hlo_passes, bool use_threads,
+    const std::optional<ErrorSpec>& error) {
+  absl::StatusOr<std::vector<Literal>> fake_arguments = MakeFakeArguments(
+      /*module=*/module_0.get(), /*pseudo_random=*/true,
+      /*use_large_range=*/false,
+      /*treat_gte_as_data_formatting=*/false,
+      /*max_bits_of_precision=*/std::nullopt);
+  CHECK_OK(fake_arguments);
+  std::vector<const Literal*> fake_argument_ptrs;
+  absl::c_transform(
+      /*input=*/*fake_arguments,
+      /*output=*/std::back_inserter(fake_argument_ptrs),
+      /*unary_op=*/[](const Literal& literal) -> Literal* {
+        return const_cast<Literal*>(&literal);
+      });
+  HloRunner::ReplicatedExecuteOptions options{
+      .num_replicas = module_0->config().replica_count(),
```

Comment:
Tensorflow hasn't switched to C++20 yet. Can you please avoid designated initializers?